askama_shared/src/i18n_runtime.rs

```diff
@@ -0,0 +1,322 @@
+//! Code used in the implementation of `impl_localize!` and the `localize()` filter.
+//!
+//! Everything in this module should be considered an internal implementation detail; it is only public
+//! for use by the macro.
+//!
+//! Maintenance note: in general, the policy is to move as much i18n code as possible into here;
+//! whatever absolutely *must* be included in the generated code is done in askama_derive.
+
+use accept_language::parse as accept_language_parse;
+use fluent_bundle::{FluentBundle, FluentResource, FluentValue};
+use std::collections::HashMap;
+
+use super::{Error, Result};
+
+pub use lazy_static::lazy_static;
+
+/// An I18n argument value. Instantiated only by the `{ localize() }` filter.
+pub type I18nValue = fluent_bundle::FluentValue;
+
+/// A known locale. Instantiated only by the `impl_localize!` macro.
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Clone, Copy)]
+pub struct Locale(pub &'static str);
+
+/// Sources; an array mapping Locales to fluent source strings. Instantiated only by the `impl_localize!` macro.
+pub type Sources = &'static [(Locale, &'static str)];
+
+/// Sources that have been parsed. Instantiated only by the `impl_localize!` macro.
+///
+/// This type is initialized in a lazy_static! in impl_localize!,
+/// because FluentBundle can only take FluentResources by reference;
+/// we have to store them somewhere to reference them.
+/// This can go away once https://github.com/projectfluent/fluent-rs/issues/103 lands.
```

Comment:
So that issue mentions some apparently newly available features/crates -- have you looked into their applicability yet? From looking over the code here, I'm mostly worried about the finer details of fallback chains. (For example, I would say that `en` should prefer `en_US` over `en_AU` in most projects -- though I'm not sure if the Fluent code handles it like this?)

Reply:
Those crates don't really buy us anything, honestly. `fluent-resmgr` is designed for loading from the filesystem, which we don't do, and `fluent-fallback` does a bunch of unnecessary allocation -- re-building parts of the localization system for every call, which isn't necessary. (It's also only [50 lines of code](https://github.com/projectfluent/fluent-rs/blob/master/fluent-fallback/src/lib.rs), so i don't think it's bad to re-implement it more efficiently.)

...On the other hand, https://crates.io/crates/fluent-locale does exactly what we want here! It wasn't used in the examples before, i didn't realize it existed. Will switch to using it now :)